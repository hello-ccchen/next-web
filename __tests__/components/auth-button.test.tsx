/**
 * @jest-environment jsdom
 */
import { fireEvent, render } from "@testing-library/react";
import { signOut, useSession } from "next-auth/react";
import { Session } from "next-auth";
import AuthButton from "@/components/auth-button";

jest.mock("next-auth/react", () => ({
  signOut: jest.fn<void, []>(() => Promise.resolve()),
  useSession: jest.fn(),
}));

describe("AuthButton component", () => {
  it("renders `Sign in` text when user's session not exists", () => {
    (useSession as jest.Mock).mockReturnValue({
      status: "unauthenticated",
      data: null,
    });

    const { container, getByText } = render(<AuthButton />);
    expect(container).toMatchSnapshot();
    expect(getByText("Sign in")).toBeTruthy();
  });

  it("renders logged-in user's info when user's session exists", () => {
    const mockSession: Session = {
      user: { name: "Test User" },
      expires: "",
    };

    (useSession as jest.Mock).mockReturnValue({
      status: "authenticated",
      data: mockSession,
    });

    const { container, getByText } = render(<AuthButton />);
    expect(container).toMatchSnapshot();
    expect(getByText("Test User")).toBeTruthy();
  });

  it("should sign out when user click the sign out button", () => {
    const mockSession: Session = {
      user: { name: "Test User" },
      expires: "",
    };

    (useSession as jest.Mock).mockReturnValue({
      status: "authenticated",
      data: mockSession,
    });

    const { getByAltText } = render(<AuthButton />);
    const signOutButton = getByAltText("logout");
    fireEvent.click(signOutButton);
    expect(signOut).toHaveBeenCalledTimes(1);
  });
});
