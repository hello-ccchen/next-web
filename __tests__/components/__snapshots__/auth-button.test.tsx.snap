// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AuthButton component renders \`Sign in\` text when user's session not exists 1`] = `
<div>
  <a
    href="/login"
  >
    Sign in
    <img
      alt="login"
      class="ms-2"
      data-nimg="1"
      decoding="async"
      height="24"
      loading="lazy"
      src="/login.svg"
      style="color: transparent;"
      width="24"
    />
  </a>
</div>
`;

exports[`AuthButton component renders logged-in user's info when user's session exists 1`] = `
<div>
  <div>
    Test User
    <img
      alt="logout"
      class="ms-2"
      data-nimg="1"
      decoding="async"
      height="24"
      loading="lazy"
      src="/logout.svg"
      style="color: transparent; cursor: pointer;"
      width="24"
    />
  </div>
</div>
`;
